### PR TITLE
Improve circular dependency handling to NessieClient

### PR DESCRIPTION
**Pull request by nastra**
The circular dependecy is fixed by introducing the same NessieClient version that Iceberg uses.

This also moves NessieClient instantiation into subclasses of AbstractSparkTest so that each subclass can control how to instantiate the NessieClient (depending on the Nessie version they pull in)

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/1783)

### CODE REVIEW

File: clients/spark-extensions/pom.xml [hunk 1]
```diff
@@ -63,7 +58,6 @@
     <dependency>
       <groupId>org.apache.iceberg</groupId>
       <artifactId>iceberg-spark3-runtime</artifactId>
-      <scope>test</scope>
```
Comment (rymurr) [inline clients/spark-extensions/pom.xml:66]: this has to be a test scope. The code has no dependency on iceberg runtime. This would ship the iceberg runtime w/ delta artifacts, which we don't want

**Review by rymurr — CHANGES_REQUESTED** [review]
I think it would be good to follow a similar pattern to the gc poms. Where the server is started up on a certain version of nessie so we are explicit about what is supported here. We have to provide some nessie client w/o explicitly adding iceberg or delta as runtime dependencies though,

**Comment by nastra** [comment]
I just opened https://github.com/projectnessie/nessie/pull/1785 to fix a similar issue with GC stuff

File: clients/spark-extensions/pom.xml (hunk 1 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml:66]: `NessieUtils` is using the `NessieClient`, therefore I removed the `test` scope so that `NessieUtils` uses whatever Nessie version that comes with `iceberg-spark3-runtime`

Comment (rymurr) [reply clients/spark-extensions/pom.xml:66]: yes, but if we do that then the iceberg runtime (all 100MB of it) becomes a hard dependency. We don't want that. This module isn't iceberg specific and we don't want to add iceberg as a dependency

**Comment by rymurr** [comment]
> I just opened #1785 to fix a similar issue with GC stuff

I think the gc stuff is correct, it doesnt have a circular dependency currently

**Comment by rymurr** [comment]
or more correctly, it manages the circular dependency by being speciifc about which versions are used in tests

**Comment by nastra** [comment]
@rymurr I pushed a new version that forces a specific version of the NessieClient on the Spark Sql extensions and the GC modules

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java [hunk 2]
```diff
@@ -49,6 +49,8 @@
 
   protected static NessieClient nessieClient;
 
+  protected abstract void initNessieClient();
```
Comment (snazy) [inline clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:52]: this should return the `NessieClient` and not let the implementations update the field in this class IMO

Comment (snazy) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:52]: the new mechanism also creates a new `NessieClient` instance for each test, but never closes them. either create only one instance or close the instances after each test (I'd prefer one instance per class)

**Comment by snazy** [comment]
Can you elaborate more on the "circular" dependeny in the PR description (and later commit summary)?

**Comment by rymurr** [comment]
would it be easier to make the abstract class not depend on a nessie client or make the dependency <optional>? Seems like an awkward way to ensure the tests run?

**Comment by nastra** [comment]
> Can you elaborate more on the "circular" dependeny in the PR description (and later commit summary)?

The problem is that the Spark SQL extentions for example rely on `iceberg-spark3-runtime`, which in turn is compiled against a Nessie version that isn't necessarily the same that we use here. Now when the tests run, we run against a `nessie-client` version that is different from Iceberg and stuff breaks. 
This happens when moving things around in the `nessie-client` project (as I'm doing in https://github.com/projectnessie/nessie/pull/1767).

**Comment by nastra** [comment]
> would it be easier to make the abstract class not depend on a nessie client or make the dependency ? Seems like an awkward way to ensure the tests run?

yes that would probably better. Let me remove the client entirely from that project

File: versioned/gc/iceberg/pom.xml [hunk 3]
```diff
@@ -30,11 +30,13 @@
...
   <name>Nessie - Versioned - GC for Spark - Iceberg</name>
 
-  <properties>
-    <iceberg.nessie.version>${project.version}</iceberg.nessie.version>
-  </properties>
-
   <dependencies>
+    <dependency>
+      <groupId>org.projectnessie</groupId>
+      <artifactId>nessie-client</artifactId>
```
Comment (rymurr) [inline versioned/gc/iceberg/pom.xml:36]: for this and above I would rather not add a nessie version explicitly. There is one in the iceberg runtime and we want to make sure the iceberg runtime one is used. As taht is how it will be used in production. This could mask issues w/ the client versions

File: clients/tests/pom.xml [hunk 4]
```diff
@@ -48,13 +48,6 @@
       <version>${spark3.version}</version>
       <scope>provided</scope>
     </dependency>
-    <dependency>
-      <groupId>org.projectnessie</groupId>
-      <artifactId>nessie-client</artifactId>
-      <version>${project.version}</version>
-      <scope>provided</scope>
```
Comment (rymurr) [inline clients/tests/pom.xml:55]: Why doesn't provided or optional work? Are the upcoming changes to the client so drastic that we have to add this code duplication? why can't the clients/tests module import the `iceberg.nessie.version` of the client instead of this refactor?

Comment (nastra) [reply clients/tests/pom.xml:55]: the idea is that delta for example might end up having a different version of the `nessie-client` than iceberg or not? That's why it seems safer to move the `nessie-client` to where it's actually required (into the subprojects)

File: versioned/gc/iceberg/pom.xml (hunk 3 above)
Comment (nastra) [reply versioned/gc/iceberg/pom.xml:36]: let me double-check whether this works without adding an explicit dependency to `nessie-client`

Comment (nastra) [reply versioned/gc/iceberg/pom.xml:36]: ah now I remember why it was failing. I missed the fact that `iceberg.nessie.version` was defined in each of the GC poms and I only updated one of them and wondered why it failed, thus leading me to adding an explicit `nessie-client` dependency

File: clients/tests/pom.xml (hunk 4 above)
Comment (nastra) [reply clients/tests/pom.xml:55]: so technically speaking we could then also have a `deltalake.nessie.version` which is being used in the delta projects, thus making our lives easier in case Delta&Iceberg stuff uses different nessie versions

File: versioned/gc/iceberg/pom.xml (hunk 3 above)
Comment (nastra) [reply versioned/gc/iceberg/pom.xml]: fixed it

Comment (nastra) [reply versioned/gc/iceberg/pom.xml]: ok i double checked this and it doesn't work without forcing the `nessie-client` version here to be `${iceberg.nessie.version}`. It's either that, or we run into the jackson issue that you already ran into (where you moved the `iceberg-spark3-runtime` dependency to the very end of the pom:
```
Caused by: org.apache.iceberg.shaded.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: 
Cannot construct instance of `org.projectnessie.model.Reference` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
```

**Review by snazy — COMMENTED** [review]
The logic to create the nessie-client, delete branches after tests and close the client is duplicated a lot. Can you move that into `AbstractSparkTest` - at best deleting all existing branches and then re-creating the default branch?
A `compileOnly` scope should work.

**Comment by nastra** [comment]
> The logic to create the nessie-client, delete branches after tests and close the client is duplicated a lot. Can you move that into `AbstractSparkTest` - at best deleting all existing branches and then re-creating the default branch?
> A `compileOnly` scope should work.

The whole idea is to leave the `nessie-client` out of `clients/tests/pom.xml` because it's possible that subprojects (delta, iceberg) at different versions might depend on different versions of `nessie-client`. Right now this is only reflected via `iceberg.nessie.version` but I imagine us also having a `delta.nessie.version` field

File: versioned/gc/iceberg-actions/pom.xml [hunk 5]
```diff
@@ -30,11 +30,12 @@
...
-  <properties>
-    <iceberg.nessie.version>${project.version}</iceberg.nessie.version>
-  </properties>
-
   <dependencies>
+    <dependency>
+      <groupId>org.projectnessie</groupId>
+      <artifactId>nessie-client</artifactId>
+      <version>${iceberg.nessie.version}</version>
+    </dependency>
```
Comment (snazy) [inline versioned/gc/iceberg-actions/pom.xml:38]: Could be 
```

    <dependency>
      <groupId>org.projectnessie</groupId>
      <artifactId>nessie-client</artifactId>
      <version>${iceberg.nessie.version}</version>
      <scope>test</scope>
      <exclusions>
        <exclusion>
          <groupId>org.projectnessie</groupId>
          <artifactId>nessie-model</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
 ```

Comment (nastra) [reply versioned/gc/iceberg-actions/pom.xml:38]: context: we discussed this during standup where @snazy had a similar problem.

@snazy do you want to fix this as part of a follow-up PR or should I fix it here?

Comment (nastra) [reply versioned/gc/iceberg-actions/pom.xml:38]: I'll make a follow up PR shortly with a concrete problem description for this part

**Review by snazy — COMMENTED** [review]
I think, all dependencies to `<artifactId>nessie-client</artifactId>` in the changed poms should have the `<scope>test</scope>`.

File: clients/spark-extensions/pom.xml [hunk 6]
```diff
@@ -33,7 +33,7 @@
     <dependency>
       <groupId>org.projectnessie</groupId>
       <artifactId>nessie-client</artifactId>
-      <version>${project.version}</version>
+      <version>${iceberg.nessie.version}</version>
```
Comment (snazy) [inline clients/spark-extensions/pom.xml:36]: Does this particular one really need to be bumped?
I'm a bit worried about this particular one, because it's a production-dependency and `nessie-client` depends on `nessie-model`.

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java [hunk 7]
```diff
@@ -67,8 +64,6 @@ protected void create() throws IOException {
         .set("spark.sql.catalog.nessie", "org.apache.iceberg.spark.SparkCatalog");
     spark = SparkSession.builder().master("local[2]").config(conf).getOrCreate();
     spark.sparkContext().setLogLevel("WARN");
-
-    nessieClient = NessieClient.builder().withUri(url).build();
   }
 
   @AfterAll
```
Comment (snazy) [inline clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:69]: Can you change this to `@AfterEach` as well? It feels weird to initialize a SparkSession for every test but only clean up the last one. Can be a follow-up though.

File: clients/spark-extensions/pom.xml (hunk 6 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml:36]: yes, this definitely needs to be bumped

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java (hunk 7 above)
Comment (nastra) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:69]: yeah should probably be a follow-up

File: clients/spark-extensions/pom.xml (hunk 6 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml:36]: you mean we should also exclude the model classes here?

Comment (nastra) [reply clients/spark-extensions/pom.xml:36]: I had another look at the pom and turns out this was running with the nessie-quarkus `project.version` and not the `iceberg.nessie.version`. Pushed a fix for this and also forced the `nessie-model` version to be `iceberg.nessie.version`

File: clients/spark-extensions/pom.xml [hunk 8]
```diff
@@ -136,10 +148,10 @@
       <plugin>
         <groupId>org.projectnessie</groupId>
         <artifactId>nessie-apprunner-maven-plugin</artifactId>
-        <version>${project.version}</version>
+        <version>${iceberg.nessie.version}</version>
```
Comment (snazy) [inline clients/spark-extensions/pom.xml:151]: Unnecessary change - the plugin can still be the project's version

File: clients/spark-extensions/pom.xml (hunk 6 above)
Comment (snazy) [reply clients/spark-extensions/pom.xml:36]: Really not happy w/ having these versions as _production_ dependencies here. I think, that makes it really complicated to reason about and prone to more difficult dependency issues.

It's already unclear, which nessie and 3rd party dependencies will land here in which versions. `nessie-client-tests` brings in more dependencies.

The changes to the gc-modules are fine IMO, because those only affect tests, but this change affects production dependencies.

File: clients/spark-extensions/pom.xml (hunk 8 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml:151]: it's necessary with the `nessie-model` changes and startup breaks

File: clients/spark-extensions/pom.xml (hunk 6 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml:36]: @rymurr thoughts on @snazy's concerns? I currently don't see any easier way around this problem than running the spark sql extensions code with the nessie version that comes with iceberg.
imo we need to rethink the dependencies between nessie and iceberg entirely as the way it is currently structured is just error prone

Comment (rymurr) [reply clients/spark-extensions/pom.xml:36]: why can't this just be `provided` or `optional` scope? We just don't ship it with the shaded jar or as a dependency and then the delta or iceberg runtimes use their nessie client that is already available to them.

I thikn we need to make sure it generates a dependency reduced pom and that shading is correct though

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java [hunk 9]
```diff
@@ -67,12 +64,10 @@ protected void create() throws IOException {
...
     spark = SparkSession.builder().master("local[2]").config(conf).getOrCreate();
     spark.sparkContext().setLogLevel("WARN");
-
-    nessieClient = NessieClient.builder().withUri(url).build();
   }
 
-  @AfterAll
-  static void tearDown() throws Exception {
+  @AfterEach
+  protected void tearDown() {
```
Comment (rymurr) [inline clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:70]: I don't think we want to stop spark after each run. Why change this?

File: versioned/gc/iceberg-actions/pom.xml [hunk 10]
```diff
@@ -30,11 +30,20 @@
 
   <name>Nessie - Versioned - GC for Spark - Iceberg - Actions</name>
 
-  <properties>
-    <iceberg.nessie.version>${project.version}</iceberg.nessie.version>
-  </properties>
-
   <dependencies>
+    <dependency>
```
Comment (rymurr) [inline versioned/gc/iceberg-actions/pom.xml:34]: why does this have to be added? When it exists in the iceberg runtime?

File: clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java [hunk 11]
```diff
@@ -44,6 +47,21 @@
 
   @TempDir File tempPath;
 
+  private NessieClient nessieClient;
```
Comment (rymurr) [inline clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java:50]: Why does this have to live here, why not in abstract test? (same for other tests)

**Review by rymurr — COMMENTED** [review]
I am not 100% sure why we cant have a single `iceberg.nessie.version`, which is renamed `client.nessie.version` that we support and assume all the things we ship to clients only work on those. Until we sort out backwards compatibility permanently (eg until 1.0). I think this change can be reduced in scope then. We are saying we only support a single version of pre-1.0 nessie at any time.

File: clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java (hunk 11 above)
Comment (nastra) [reply clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java:50]: see my comment in https://github.com/projectnessie/nessie/pull/1783/files#r691250922

File: versioned/gc/iceberg-actions/pom.xml (hunk 10 above)
Comment (nastra) [reply versioned/gc/iceberg-actions/pom.xml]: outdated code

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java (hunk 9 above)
Comment (nastra) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:70]: requested in https://github.com/projectnessie/nessie/pull/1783#discussion_r693755112

Comment (rymurr) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:70]: Spark initialization is a bit tricky and stopping it like this could lead to funky issues. It maintains a lot of static state. It also takes a long time to start up so this adds a ton of time to tests.

@snazy we aren't initializing the Spark server in every run, we are initializing it in the first then getting it in the rest (hence `getOrCreate`). If this is particularly bothersome i would `create` in `BeforeAll` and `get` in `BeforeEach`

File: clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java (hunk 11 above)
Comment (rymurr) [reply clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java:50]: yes, I don't think that answered the question. As per my [comment](https://github.com/projectnessie/nessie/pull/1783#pullrequestreview-736888401) I don't see what is wrong w forcing 1 (pre1.0) client version for all these tests and using it from the abstract test. The solution here creates a lot of duplicate code and confusing pom changes imho

Comment (nastra) [reply clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java:50]: but won't we eventually end up having different nessie-client versions for Delta and Iceberg? How do we guarantee that all the things use the same `client.nessie.version` when Delta/Iceberg release at a different pace?

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java (hunk 9 above)
Comment (snazy) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java:70]: Bummer - missed the `getOrCreate`

File: clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java (hunk 11 above)
Comment (rymurr) [reply clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java:50]: my thinking is the following:

1. we only need this pre-1.0 as we will be introducing backwards compatibility soon
2. the delta code will soon either be deleted or moved to the delta repo depending on discussions w/ Databricks.

**Comment by snazy** [comment]
I doubt we need this change with #1816, because that abstracts changes to Nessie-Client and its implementation(s) from the call sites.

**Comment by rymurr** [comment]
> I doubt we need this change with #1816, because that abstracts changes to Nessie-Client and its implementation(s) from the call sites.

can you expand on that...i am not sure what you mean.

File: clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java (hunk 11 above)
Comment (nastra) [reply clients/deltalake/core/src/test/java/org/projectnessie/deltalake/ITDeltaLog.java]: ok seems reasonable. Pushed a new simplified version and renamed `iceberg.nessie.version` to `client.nessie.version`

**Comment by snazy** [comment]
> > I doubt we need this change with #1816, because that abstracts changes to Nessie-Client and its implementation(s) from the call sites.
> 
> can you expand on that...i am not sure what you mean.

AFAIK this PR and #1818 are "caused by" #1767, which changes `NessieClient` and its builders (breaks backwards compatibility). #1816 introduces client-builder interfaces but keeps the existing `NessieClient.Builder` for backwards-compatibility (can be removed once all projects that depend on it - aka Iceberg) have been migrated.

**Comment by nastra** [comment]
> > > I doubt we need this change with #1816, because that abstracts changes to Nessie-Client and its implementation(s) from the call sites.
> > 
> > 
> > can you expand on that...i am not sure what you mean.
> 
> AFAIK this PR and #1818 are "caused by" #1767, which changes `NessieClient` and its builders (breaks backwards compatibility). #1816 introduces client-builder interfaces but keeps the existing `NessieClient.Builder` for backwards-compatibility (can be removed once all projects that depend on it - aka Iceberg) have been migrated.

Right, but #1816 only overcomes the problem for the HttpClient stuff. Next time when somebody changes something in `nessie-client` before we have backwards compatibility in place, that person runs into the same problem as we have right now

**Comment by snazy** [comment]
> Right, but #1816 only overcomes the problem for the HttpClient stuff. Next time when somebody changes something in `nessie-client` before we have backwards compatibility in place, that person runs into the same problem as we have right now

Maybe, maybe not. Any change post GA must be implemented in a backwards compatible way w/ a migration strategy.

File: clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java (hunk 9 above)
Comment (nastra) [reply clients/tests/src/main/java/org/projectnessie/client/tests/AbstractSparkTest.java]: reverted those changes

File: clients/spark-extensions/pom.xml (hunk 6 above)
Comment (nastra) [reply clients/spark-extensions/pom.xml]: fixed

**Review by rymurr — APPROVED** [review]
